woocommerce/woocommerce
# Add support to send files in Slack CLI tool

### Changes proposed in this Pull Request:

@jonathansadowski wanted to introduce file sending, so here is a proposed approach. Looking at the old command I built it at the time not realizing that @slack/web-api existed. This gets around a bunch of issues with extending the old approach too (such as our use of Node 16 which doesn't have built in FormData without enabling experimental features and the `form-data` lib has security issues).

As you can see with this approach its much simpler anyway. I felt it's most simple to introduce a separate command here mainly because each endpoint accepts different optional args and also has different required scope, so i think its best to be explicit.

Note there are some caveats to this approach to because:

1. for file upload the bot must be a member of the channel, that's not required for the sending message approach we take
2. We could continue using channel names instead of IDs but that means the bot needs an extra scope to list out all the channels then find the channel ID. Happy to do that but I think maybe just using IDs directly is better for now.
3. You'll also need your bot to have `files:write` permission. Please note if you change your existing bot **you will need to reinstall it in the slack group, slack provides a convenience link to this when you change a permission.**



### How to test the changes in this Pull Request:

1. Follow the first steps of https://github.com/woocommerce/woocommerce/pull/38185 to setup a bot and token and slack group for testing

2. Ensure your bot also has `files:write` permission and is also a member of the channel you want to send the file to. (`/invite @bot_name` in the channel).

3. Get the channel ID from slack: https://stackoverflow.com/a/76536392/472987

5. Now run the command on this branch from root:

`pnpm utils slack file "your-app-oauth-token" "Your message" "./path/to/your/file.txt" "channel ID"`




### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
